src/transformers/models/mistral/modeling_mistral.py

```diff
@@ -1076,6 +1029,131 @@ def forward(
+
+        if self.config.sliding_window is not None:
+            if attention_mask is not None and attention_mask.dim() == 4:
+                logger.warning_once(
+                    "Sliding window will not take effect when passing 4d custom masks"
+                    "you may get unexpected results, use attention mask generated by tokenizer"
+                    "or set model.config.sliding_window to None if you don't want sliding window"
+                )
+
+            #  can only happen in prefill phase, when the prompt length > sliding window length, we need to do this
+            #  manually because we are returning the whole prompt token sequence in `SlidingWindowCache`, maybe a better
+            #  way is to support chunked prefill instead
+            if sequence_length > self.config.sliding_window and using_sliding_window_cache:
+                exclude_mask |= torch.arange(target_length, device=device) <= (
+                    cache_position.reshape(-1, 1) - self.config.sliding_window
+                )
+
+            # not using `SlidingWindowCache` and attention mask supports sliding window
+            if (attention_mask is None or attention_mask.dim() == 2) and not using_sliding_window_cache:
+                exclude_mask |= torch.arange(target_length, device=device) <= (
+                    cache_position.reshape(-1, 1) - self.config.sliding_window
+                )
+
+        causal_mask *= exclude_mask
+        causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
+        if attention_mask is not None:
+            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+            if attention_mask.dim() == 2:
+                mask_length = attention_mask.shape[-1]
+                padding_mask = causal_mask[..., :mask_length].eq(0.0) * attention_mask[:, None, None, :].eq(0.0)
+                causal_mask[..., :mask_length] = causal_mask[..., :mask_length].masked_fill(padding_mask, min_dtype)
+            elif attention_mask.dim() == 4:
+                # backwards compatibility: we allow passing a 4D attention mask shorter than the input length with
+                # cache. In that case, the 4D attention mask attends to the newest tokens only.
+                if attention_mask.shape[-2] < cache_position[0] + sequence_length:
+                    logger.warning_once(
+                        "Passing a 4d mask shorter than the input lenght is deprecated and will be "
+                        "removed in transformers v4.42.0"
+                    )
+                    offset = cache_position[0]
+                else:
+                    offset = 0
+                mask_shape = attention_mask.shape
+                mask_slice = (attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype
+                causal_mask[
+                    : mask_shape[0], : mask_shape[1], offset : mask_shape[2] + offset, : mask_shape[3]
+                ] = mask_slice
```

Comment:
no here let's just return the attention mask (we updated that in llama)

Reply:
The current main branch llama still have the elif branch though